# **WIE3007: Data Mining and Warehousing**
# **Group Project**
## **Group 11**
### 'Ammar Haziq bin Abdul Halim (22002548)
### Muhammad Nur Fitri bin Faizan (22001866)
### Shaarani Navaratnam (23005234)
### Nur Dhiya Syafia binti Mohd Noor Salizan (22002096)
### Silvia Evafarina Asun (23062999)

# **Dataset Simulation**

In [1]:
pip install pandas numpy matplotlib seaborn scikit-learn python-dotenv

  Using cached pytz-2025.2-py2.py3-none-any.whl.metadata (22 kB)
  Using cached cycler-0.12.1-py3-none-any.whl.metadata (3.8 kB)
   ---------------------------------------- 0.0/11.0 MB ? eta -:--:--
   ------ --------------------------------- 1.8/11.0 MB 9.1 MB/s eta 0:00:02
   --------------- ------------------------ 4.2/11.0 MB 9.9 MB/s eta 0:00:01
   ------------------------ --------------- 6.8/11.0 MB 10.7 MB/s eta 0:00:01
   ----------------------------------- ---- 9.7/11.0 MB 11.3 MB/s eta 0:00:01
   ---------------------------------------- 11.0/11.0 MB 10.8 MB/s eta 0:00:00
   ---------------------------------------- 0.0/12.3 MB ? eta -:--:--
   --------- ------------------------------ 2.9/12.3 MB 16.7 MB/s eta 0:00:01
   -------------------- ------------------- 6.3/12.3 MB 15.7 MB/s eta 0:00:01
   ----------------------------- ---------- 9.2/12.3 MB 14.9 MB/s eta 0:00:01
   ---------------------------------------  12.1/12.3 MB 14.7 MB/s eta 0:00:01
   --------------------------


[notice] A new release of pip is available: 25.1.1 -> 25.3
[notice] To update, run: python.exe -m pip install --upgrade pip


In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
pip install -U google-genai

  Using cached tenacity-9.1.2-py3-none-any.whl.metadata (1.2 kB)
  Using cached cachetools-6.2.4-py3-none-any.whl.metadata (5.6 kB)
Using cached cachetools-6.2.4-py3-none-any.whl (11 kB)
   ---------------------------------------- 0.0/2.0 MB ? eta -:--:--
   ---------------------------------------- 2.0/2.0 MB 11.0 MB/s eta 0:00:00
Using cached tenacity-9.1.2-py3-none-any.whl (28 kB)

   ----------------------------------------  0/24 [websockets]
   ----------------------------------------  0/24 [websockets]
   - --------------------------------------  1/24 [urllib3]
   - --------------------------------------  1/24 [urllib3]
   ----- ----------------------------------  3/24 [tenacity]
   -------- -------------------------------  5/24 [pyasn1]
   ---------- -----------------------------  6/24 [idna]
   ---------- -----------------------------  6/24 [idna]
   --------------- ------------------------  9/24 [charset_normalizer]
   ------------------ --------------------- 11/24 [cachetools]


[notice] A new release of pip is available: 25.1.1 -> 25.3
[notice] To update, run: python.exe -m pip install --upgrade pip


In [4]:
import os
from dotenv import load_dotenv
from google import genai

# Load environment variables
load_dotenv(override=True)

# Get the Gemini client
# FOR OTHER TEAM MEMBERS: If you would like to try, use your own Gemini API key
client = genai.Client(api_key=os.getenv("GEMINI_API_KEY"))

ValueError: Missing key inputs argument! To use the Google AI API, provide (`api_key`) arguments. To use the Google Cloud API, provide (`vertexai`, `project` & `location`) arguments.

In [ ]:
dataset_simulation_prompt = """
Act as a financial data simulator.
Generate 5000 rows of synthetic bank loan application data in valid CSV format.

The dataset must be realistic and contain financial patterns suitable for predictive modelling (classification).
Include the following columns:
1. loan_id: Unique identifier (eg 1001, L1002...).
2. age: Numeric (eg 18 to 70).
3. gender: Categorical ('Male', 'Female').
4. job_title: Specific job role (eg "Software Engineer", "Nurse", "Barista", "Accountant", "Driver"). Vary these significantly.
5. annual_income: Numeric (eg 20000 to 150000).
6. loan_amount: Numeric (eg 2000 to 50000).
7. loan_purpose: Categorical ('Home', 'Business', 'Education', 'Medical', 'Personal').
8. credit_score: Numeric (eg 300 to 850).
9. number_of_dependents: Numeric (0 to 5).
10. applicant_statement: A realistic text field (2 to 3 sentences) written by the applicant explaining why they need the loan and how they plan to pay it back. Vary the tone of applicant_statement to sound either confident, desperate, professional, or vague to allow for later NLP sentiment and risk analysis.
11. loan_status: Target variable (0 for Rejected/Default, 1 for Approved/Paid).

Crucial requirement for logic:
- PLEASE ensure correlations exist between the features (with realistic, logical sense) to prevent the model from being random.
- Logic 1 (Risk): Applicants with low income, high loan_amount, low credit_score, or "desperate"-sounding statements should have a higher chance of being '0' (Rejected), and whatever is common sense within this logic.
- Logic 2 (Stability): Applicants with professional job titles (eg "Engineer", "Doctor", within similar types), high credit scores, and "confident"-sounding statements have a higher chance of being '1' (Approved), and whatever is common sense within this logic.
- Logic 3 (Life Stage): Younger applicants (age under 25) should typically have lower income and shorter credit history (lower scores), and whatever is common sense within this logic.

Output ONLY the CSV data (with header, delimiter is ','), no other text.
"""

In [ ]:
# dataset_response = client.models.generate_content(
#     model="gemini-flash-lite-latest", contents=dataset_simulation_prompt
# )
#
# dataset_text = dataset_response.text

In [ ]:
# print(dataset_text)

In [ ]:
import io

# Read the dataset text into a DataFrame
# df = pd.read_csv(io.StringIO(dataset_text))

In [ ]:
# df

In [ ]:
# df.shape

In [ ]:
# df.to_csv('WIE3007_initial_dataset.csv', index=False)

In [ ]:
import pandas as pd

# To load this dataset, place 'WIE3007_initial_dataset.csv' in the same level as the .ipynb notebook, and run this
df = pd.read_csv('WIE3007_initial_dataset.csv')

# **Feature Engineering with Gemini LLM**

## **Gemini LLM Batch Prompting Functions**

In [ ]:
import json
from google.genai import types

MODEL = "gemini-2.0-flash-lite" # Fast model with high throughput

In [ ]:
def prepare_batch_prompts(df, input_file_name: str, prompt: str, column: str = None):
    """
    :param df: The DataFrame involved in the batch prompting.
    :param input_file_name: The name of the input prompts file. Recommended to be descriptive of the task to be done. Must include .jsonl file type in file name.
    :param prompt: The prompt to insert for each row.
    :param column: The column of the DataFrame that is involved in the prompt. If left empty (and has the default value of None), the whole DataFrame's columns and its values from the parameter df is used as context in the prompt.
    :return:
    """

    # 1. Build JSONL data for Gemini's Batch API
    jsonl_data = []

    # If column is specified, only df[column] is inserted as context into the prompt
    if column is not None:
        for idx, text in df[column].items():
            # The prompt is the same for every row
            final_prompt = f"{prompt}\nText: \"{text}\""

            jsonl_data.append(json.dumps({
                # Use the DataFrame index as the key to map results back later
                "key": f"index_{idx}",
                "request": {
                    "contents": [{
                        "parts": [{"text": final_prompt}],
                        "role": "user"
                    }]
                }
            }))
    else:
        # The whole row of the parameter df is taken inserted as context into the prompt
        for idx, row in df.iterrows():
            # Pass the row as a dictionary
            row_data = row.to_dict()

            # The prompt is the same for every row
            final_prompt = f"{prompt}\nContext: \"{row_data}\""

            jsonl_data.append(json.dumps({
                # Use the DataFrame index as the key to map results back later
                "key": f"index_{idx}",
                "request": {
                    "contents": [{
                        "parts": [{"text": final_prompt}],
                        "role": "user"
                    }]
                }
            }))

    # 2. Save to a temporary JSONL file
    with open(input_file_name, "w") as file:
        file.write("\n".join(jsonl_data))

    print(f"Prepared {len(df.index)} prompts in {input_file_name}.")

In [ ]:
def prepare_batch_prompts(df, column: str, input_file_name: str, prompt: str):
    """
    :param df: The DataFrame involved in the batch prompting.
    :param column: The column of the DataFrame that is involved in the prompt.
    :param input_file_name: The name of the input prompts file. Recommended to be descriptive of the task to be done. Must include .jsonl file type in file name.
    :param prompt: The prompt to insert for each row.
    :return:
    """

    # 1. Build JSONL data for Gemini's Batch API
    jsonl_data = []

    for idx, text in df[column].items():
        # The prompt is the same for every row
        final_prompt = f"{prompt}\nText: \"{text}\""

        jsonl_data.append(json.dumps({
            # Use the DataFrame index as the key to map results back later
            "key": f"index_{idx}",
            "request": {
                "contents": [{
                    "parts": [{"text": final_prompt}],
                    "role": "user"
                }]
            }
        }))

    # 2. Save to a temporary JSONL file
    with open(input_file_name, "w") as file:
        file.write("\n".join(jsonl_data))

    print(f"Prepared {len(df.index)} prompts in {input_file_name}.")

In [ ]:
def upload_batch_prompts(input_file_name: str, display_name: str):
    """
    :param input_file_name: The name of the input prompts file. Must include .jsonl file type in file name.
    :param display_name: The name to give the batch job. It should be descriptive of what the job is doing.
    :return:
    """

    # Upload the file to the Gemini API File service
    print(f"Uploading {input_file_name}...")

    global uploaded_file
    uploaded_file = client.files.upload(
        file=input_file_name,
        config=types.UploadFileConfig(display_name=display_name, mime_type='jsonl')
    )

    print(f"File uploaded successfully: {uploaded_file.name}")

    # Create the Batch Job
    print("Creating batch job...")

    global feedback_batch_job
    feedback_batch_job = client.batches.create(
        model=MODEL,
        src=uploaded_file.name, # Source is the uploaded file name
        config={'display_name': display_name}
    )

    global job_name
    job_name = feedback_batch_job.name

    print(f"Batch job created: {job_name}. Status: {feedback_batch_job.state.name}")

In [ ]:
import time

def track_batch_job(df, engineered_feature_name: str):
    """
    :param df: The DataFrame involved in the batch prompting.
    :param engineered_feature_name: The specific name of the new feature to be engineered. This will be used as the column name in the DataFrame specified in the parameter df.
    :return:
    """

    # Poll for job completion
    print("Waiting for batch job to complete (this can take several minutes or longer)...")

    global feedback_batch_job
    global job_name
    while feedback_batch_job.state.name not in ('JOB_STATE_SUCCEEDED', 'JOB_STATE_FAILED'):
        feedback_batch_job = client.batches.get(name=job_name)
        print(f"Job Status: {feedback_batch_job.state.name}. Waiting...")
        time.sleep(15) # Check every 15 seconds

    if feedback_batch_job.state.name == 'JOB_STATE_FAILED':
        raise Exception(f"Batch Job Failed: {feedback_batch_job.state.name}")

    # Job Succeeded: Retrieve the output file reference
    result_file_name = feedback_batch_job.dest.file_name
    print(f"Job Succeeded! Downloading results from: {result_file_name}")

    # Download the output file content
    file_content_bytes = client.files.download(file=result_file_name)
    file_content = file_content_bytes.decode('utf-8')

    # Process the JSONL output
    results = [json.loads(line) for line in file_content.splitlines()]

    # Map the results back to the DataFrame
    output_map = {}
    for result in results:
        key = result.get('key') # e.g., 'idx_123'
        if 'response' in result and 'candidates' in result['response']:
            # Extract the classification text
            output_text = result['response']['candidates'][0]['content']['parts'][0]['text']

            # Extract the original DataFrame index from the key
            original_idx = int(key.split('_')[1])
            output_map[original_idx] = output_text

    # Put the results back into the DataFrame using the index map
    df[engineered_feature_name] = df.index.map(output_map)

    print(f"\"{engineered_feature_name}\" feature generation complete and the DataFrame has been updated with those new values.")

## **Sentiment Engineering**

In [5]:
df_engineered = df.copy()

NameError: name 'df' is not defined

In [ ]:
sentiment_input_file_name = "sentiment_input.jsonl"
sentiment_output_file_name = "sentiment_output.jsonl"
sentiment_display_name = "sentiment_generation"

sentiment_prompt = """
Classify this loan application's applicant statement as one word of either positive, neutral, or negative ONLY. The final output should be ONE WORD ONLY and should have NO surrounding quotes, NO \\n, only the classification.
"""

In [ ]:
prepare_batch_prompts(df_engineered, 'applicant_statement', sentiment_input_file_name, sentiment_prompt)

In [ ]:
upload_batch_prompts(sentiment_input_file_name, sentiment_display_name)

In [ ]:
track_batch_job(df_engineered, 'sentiment')

In [ ]:
df_engineered

In [ ]:
df_engineered['sentiment'] = df_engineered['sentiment'].str.rstrip('\n')

In [ ]:
df_engineered

In [ ]:
df_engineered.to_csv('WIE3007_latest_dataset.csv', index=False)

## **Risk Categorisation Engineering**

In [6]:
df = pd.read_csv('WIE3007_latest_dataset.csv')

In [8]:
#checking before starting
print("Rows, Cols:", df.shape)
print("Columns:", df.columns.tolist())

# Ensure numeric columns are numeric (in case they were read as strings)
num_cols = ["annual_income", "loan_amount", "credit_score", "number_of_dependents"]
for c in num_cols:
    if c in df.columns:
        df[c] = pd.to_numeric(df[c], errors="coerce")

# Optional: age
if "age" in df.columns:
    df["age"] = pd.to_numeric(df["age"], errors="coerce")

df[num_cols].isna().sum()


Rows, Cols: (1279, 12)
Columns: ['loan_id', 'age', 'gender', 'job_title', 'annual_income', 'loan_amount', 'loan_purpose', 'credit_score', 'number_of_dependents', 'applicant_statement', 'loan_status', 'sentiment']


annual_income           0
loan_amount             0
credit_score            0
number_of_dependents    0
dtype: int64

In [9]:
#create core risk features- lon to income and buffers
# Avoid division by zero
df["annual_income_safe"] = df["annual_income"].replace(0, np.nan)

# Core financial ratio
df["loan_to_income_ratio"] = df["loan_amount"] / df["annual_income_safe"]

# Fill if income missing (fallback to median ratio)
df["loan_to_income_ratio"] = df["loan_to_income_ratio"].fillna(df["loan_to_income_ratio"].median())

# Dependents buckets (optional but helpful)
df["dependents_bucket"] = pd.cut(
    df["number_of_dependents"],
    bins=[-1, 0, 2, 10],
    labels=["0", "1-2", "3+"]
).astype(str)

df[["loan_to_income_ratio", "dependents_bucket"]].head()


,loan_to_income_ratio,dependents_bucket
0,0.176471,1-2
1,0.156250,1-2
2,0.280000,0
3,0.138889,3+
4,0.195122,1-2


In [10]:
# convert sentiment into risk signal -simple
# Sentiment risk mapping (negative => slightly higher risk)
sentiment_map = {"Positive": 0, "Neutral": 1, "Negative": 2}

if "sentiment" in df.columns:
    df["sentiment_risk"] = df["sentiment"].map(sentiment_map).fillna(1).astype(int)
else:
    df["sentiment_risk"] = 1  # default neutral if missing

df["sentiment_risk"].value_counts(dropna=False)


sentiment_risk
0    802
2    336
1    141
Name: count, dtype: int64

In [11]:
# risk score-transparent rule based scoring
#This creates an interpretable risk score using 4–5 signals:
#Credit score (main risk indicator)
#Loan-to-income ratio (affordability)
#Dependents (higher burden)
#Sentiment (soft indicator)
#Loan purpose (optional small weight)
# 1) Credit score risk (0 = best, 3 = worst)
def credit_risk(cs):
    if pd.isna(cs):
        return 2
    if cs >= 750:
        return 0
    elif cs >= 700:
        return 1
    elif cs >= 650:
        return 2
    else:
        return 3

df["credit_risk"] = df["credit_score"].apply(credit_risk)

# 2) Loan-to-income risk
def lti_risk(x):
    if pd.isna(x):
        return 2
    if x <= 0.2:
        return 0
    elif x <= 0.4:
        return 1
    elif x <= 0.6:
        return 2
    else:
        return 3

df["lti_risk"] = df["loan_to_income_ratio"].apply(lti_risk)

# 3) Dependents risk
def dep_risk(d):
    if pd.isna(d):
        return 1
    if d == 0:
        return 0
    elif d <= 2:
        return 1
    else:
        return 2

df["dependents_risk"] = df["number_of_dependents"].apply(dep_risk)

# 4) Loan purpose risk (small weight, OPTIONAL)
# (You can edit these if your group agreed on different assumptions)
purpose_risk_map = {
    "Home": 0,
    "Education": 1,
    "Business": 1,
    "Personal": 2,
    "Medical": 1
}
if "loan_purpose" in df.columns:
    df["purpose_risk"] = df["loan_purpose"].map(purpose_risk_map).fillna(1).astype(int)
else:
    df["purpose_risk"] = 1

# Final risk score (weights chosen to be explainable)
df["risk_score"] = (
    (df["credit_risk"] * 3) +      # credit matters most
    (df["lti_risk"] * 2) +         # affordability second
    (df["dependents_risk"] * 1) +  # smaller influence
    (df["sentiment_risk"] * 1) +   # soft signal
    (df["purpose_risk"] * 1)       # small signal
)

df[["credit_risk","lti_risk","dependents_risk","sentiment_risk","purpose_risk","risk_score"]].head()


,credit_risk,lti_risk,dependents_risk,sentiment_risk,purpose_risk,risk_score
0,1,0,1,0,0,4
1,3,0,1,2,2,14
2,0,1,0,0,1,3
3,1,0,2,0,1,6
4,3,0,1,0,2,12


In [12]:
#convert score to risk category - low/medium/high
def risk_category(score):
    if score <= 4:
        return "Low"
    elif score <= 8:
        return "Medium"
    else:
        return "High"

df["risk_category"] = df["risk_score"].apply(risk_category)

df["risk_category"].value_counts()


risk_category
Low       573
High      451
Medium    255
Name: count, dtype: int64

In [13]:
#check
# Show 5 examples from each category
sample = df.groupby("risk_category").apply(lambda x: x.sample(min(5, len(x)), random_state=42)).reset_index(drop=True)

cols_to_show = ["loan_id","annual_income","loan_amount","credit_score","loan_to_income_ratio",
                "number_of_dependents","sentiment","risk_score","risk_category"]

# Only show columns that exist
cols_to_show = [c for c in cols_to_show if c in df.columns]
sample[cols_to_show]


C:\Users\dhiya\AppData\Local\Temp\ipykernel_19280\4099204191.py:3: FutureWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  sample = df.groupby("risk_category").apply(lambda x: x.sample(min(5, len(x)), random_state=42)).reset_index(drop=True)


,loan_id,annual_income,loan_amount,credit_score,loan_to_income_ratio,number_of_dependents,sentiment,risk_score,risk_category
0,L1806,46000,11500,635,0.250000,0,Negative,14,High
1,L1097,50000,12000,620,0.240000,3,Negative,17,High
2,L1918,36000,8000,615,0.222222,0,Negative,14,High
3,L2115,48000,13000,640,0.270833,2,Positive,14,High
4,L1387,80000,20000,665,0.250000,3,Negative,12,High
5,L2159,165000,56000,838,0.339394,1,Positive,4,Low
6,L1204,90000,20000,750,0.222222,1,Positive,4,Low
7,L1362,100000,25000,770,0.250000,1,Positive,4,Low
8,L1947,135000,40000,805,0.296296,1,Positive,4,Low
9,L2219,92000,17000,830,0.184783,1,Positive,2,Low


In [14]:
#save the latest dataset
df.to_csv("WIE3007_latest_dataset_after_risk.csv", index=False)
print("Saved updated dataset with risk features -> WIE3007_latest_dataset_after_risk.csv")

Saved updated dataset with risk features -> WIE3007_latest_dataset_after_risk.csv


## **Customer Segmentation Engineering**

# **Data Preprocessing**

In [ ]:
df = pd.read_csv('WIE3007_latest_dataset.csv')

In [ ]:
df

In [ ]:
df.dtypes

In [ ]:
from sklearn.preprocessing import LabelEncoder

df_preprocessed = df.copy()

le = LabelEncoder()

categorical_columns = ['gender', 'job_title', 'loan_purpose', 'sentiment']

for column in categorical_columns:
    df_preprocessed[column] = le.fit_transform(df_preprocessed[column])

In [ ]:
from sklearn.model_selection import train_test_split

X = df_preprocessed.drop(columns=['loan_id', 'applicant_statement', 'loan_status'])
y = df['loan_status']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=22002548)

In [ ]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

scaler.fit(X_train)

X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)

# **Predictive Model Development**

Show examples of underfitting and overfitting, and how to overcome those scenarios. Compare your model with at least two models (is the model performance  underfitting/overfitting/close fit compared to at least two other models).

## **Base Model: Decision Tree (DT)**

In [ ]:
from sklearn.tree import DecisionTreeClassifier

base_dtc = DecisionTreeClassifier(random_state=22002548)

base_dtc.fit(X_train_scaled, y_train)

In [ ]:
y_pred_base_dtc_train = base_dtc.predict(X_train_scaled)

y_prob_base_dtc_train = base_dtc.predict_proba(X_train_scaled)[:, 1]

y_pred_base_dtc_test = base_dtc.predict(X_test_scaled)

y_prob_base_dtc_test = base_dtc.predict_proba(X_test_scaled)[:, 1]

## **Bagging Model (DT with Bagging)**

## **Logistic Regression (LR)**

## **Random Forest (RF)**

## **Gradient Boosting (GBoost)**

## **Extreme Gradient Boosting (XGBoost)**

## **Light Gradient Boosting Machine (LightGBM)**

## **Neural Network (NN)**

# **Model Evaluation & Interpretation**

Compare models with the base model, Decision Tree, to compare how bagging, boosting, or different approaches (neural network) affect model performance with respect to a base decision tree using accuracy, precision, recall, (macro) F1 score, and ROC-AUC. Use Gemini LLM to summarise findings in the model, provide insights, and interpret feature importance.

In [ ]:
from sklearn.metrics import accuracy_score, f1_score, roc_auc_score, classification_report, confusion_matrix, ConfusionMatrixDisplay

def evaluate_model(model, y_train, y_test, y_pred_train, y_pred_test, y_prob_train, y_prob_test, model_name='Model'):
    """
    Evaluates a classification model, with comparisons between training and testing sets to check for overfitting, and analysing the classification metrics and confusion matrix.

    :param model: The model to be evaluated.
    :param y_train: The training set of the target variable.
    :param y_test: The testing set of the target variable.
    :param y_pred_train: The predictions of the target variable made using the training set.
    :param y_pred_test: The predictions of the target variable made using the testing set.
    :param y_prob_train: The probabilities of the prediction of the target variable made using the training set (output of predict_proba()).
    :param y_prob_test: The probabilities of the prediction of the target variable made using the testing set (output of predict_proba()).
    :param model_name: The name of the model being evaluated. Defaults to "Model".
    :return:
    """

    print(f"{model_name} Evaluation (Training Set)")
    print(f"Accuracy: {accuracy_score(y_train, y_pred_train):.4f}")
    print(f"F1 score: {f1_score(y_train, y_pred_train):.4f}")
    print(f"ROC-AUC: {roc_auc_score(y_train, y_prob_train):.4f}")
    print("\nClassification Report:\n", classification_report(y_train, y_pred_train))

    cm_train = confusion_matrix(y_train, y_pred_train, labels=model.classes_)
    cm_train_display = ConfusionMatrixDisplay(confusion_matrix=cm_train, display_labels=model.classes_)

    cm_train_display.plot()

    plt.show()

    print(f"{model_name} Evaluation (Testing Set)")
    print(f"Accuracy: {accuracy_score(y_test, y_pred_test):.4f}")
    print(f"F1 score: {f1_score(y_test, y_pred_test):.4f}")
    print(f"ROC-AUC: {roc_auc_score(y_test, y_prob_test):.4f}")
    print("\nClassification Report:\n", classification_report(y_test, y_pred_test))

    cm_test = confusion_matrix(y_test, y_pred_test, labels=model.classes_)
    cm_test_display = ConfusionMatrixDisplay(confusion_matrix=cm_test, display_labels=model.classes_)

    cm_test_display.plot()

    plt.show()

In [ ]:
def summarise_model_findings(model, y_train, y_test, y_pred_train, y_pred_test, y_prob_train, y_prob_test, model_name='Model'):
    """
    Uses the results of model evaluation (from evaluate_model()) to place them in a string to be sent to a Gemini LLM for further summarisation and insights to relevant stakeholders of the model.

    :param model: The model to be evaluated.
    :param y_train: The training set of the target variable.
    :param y_test: The testing set of the target variable.
    :param y_pred_train: The predictions of the target variable made using the training set.
    :param y_pred_test: The predictions of the target variable made using the testing set.
    :param y_prob_train: The probabilities of the prediction of the target variable made using the training set (output of predict_proba()).
    :param y_prob_test: The probabilities of the prediction of the target variable made using the testing set (output of predict_proba()).
    :param model_name: The name of the model being evaluated. Defaults to "Model".
    :return:
    """

    # Get feature importances of model (applies for decision tree-type models)
    if hasattr(model, 'feature_importances_'):
        feature_importances = model.feature_importances_
        feature_names = X_train.columns

        # Get the feature importance
        feature_importances_df = pd.DataFrame({'Feature': feature_names, 'Importance': feature_importances})

        # Sort the feature importances by magnitude of feature importance
        feature_importances_df = feature_importances_df.sort_values(by='Importance', ascending=False)

        # Reset index of df
        feature_importances_df = feature_importances_df.reset_index()
    else:
        feature_importances_df = "Not a decision tree-type model so has no feature importances"

    # Determine whether the feature importance is present or not
    if isinstance(feature_importances_df, pd.DataFrame):
        final_feature_importances_string = feature_importances_df.to_string(index=False)
    else:
        final_feature_importances_string = feature_importances_df

    # Set the model findings prompt
    summarise_findings_prompt = f"""
    {model_name} Findings (Training Set)
    Accuracy: {accuracy_score(y_train, y_pred_train):.4f}
    F1 score: {f1_score(y_train, y_pred_train):.4f}
    ROC-AUC: {roc_auc_score(y_train, y_prob_train):.4f}
    \nClassification Report:\n{classification_report(y_train, y_pred_train)}
    \nConfusion Matrix\n{confusion_matrix(y_train, y_pred_train, labels=model.classes_)}

    \n\n{model_name} Findings (Testing Set)
    Accuracy: {accuracy_score(y_test, y_pred_test):.4f}
    F1 score: {f1_score(y_test, y_pred_test):.4f}
    ROC-AUC: {roc_auc_score(y_test, y_prob_test):.4f}
    \nClassification Report:\n{classification_report(y_test, y_pred_test)}
    \nConfusion Matrix\n{confusion_matrix(y_test, y_pred_test, labels=model.classes_)}

    \n\n{model_name} Feature Importances Data:\n
    {final_feature_importances_string}\n\n

    You are a financial data analyst. Please summarise these findings of the {model_name} bank loan approval model and provides deep, thoughtful, elaborate insights for it. Ensure to interpret the feature importances for the model as well, explaining it. Your summarisation has to be understandable and useful to the appropriate stakeholders of this model.
    """

    summarised_findings = client.models.generate_content(
        model="gemini-3-pro-preview",
        contents=summarise_findings_prompt
    )

    print(summarised_findings.text)

## **Base Model: Decision Tree (DT)**

In [ ]:
evaluate_model(base_dtc, y_train, y_test, y_pred_base_dtc_train, y_pred_base_dtc_test, y_prob_base_dtc_train, y_prob_base_dtc_test, model_name='Base DecisionTreeClassifier')

In [ ]:
summarise_model_findings(base_dtc, y_train, y_test, y_pred_base_dtc_train, y_pred_base_dtc_test, y_prob_base_dtc_train, y_prob_base_dtc_test, model_name='Base DecisionTreeClassifier')

Output of Gemini LLM of summarised findings:
### 1. Executive Summary
The Base Decision Tree Classifier has demonstrated **exceptional predictive capability** in automating loan approval decisions. The model achieves a 95.7% accuracy rate on unseen data, indicating it is highly reliable for deployment.

However, the analysis reveals a heavy reliance on a single metric—**Credit Score (77.6% importance)**—to make these decisions. While this aligns with traditional risk assessment, the "perfect" performance on the training data suggests the model is slightly overfitted (too complex), effectively memorizing historical applications rather than purely learning patterns. Despite this, the model remains robust and safe for operational use, provided specific compliance checks regarding demographic features are conducted.

---

### 2. Model Performance Analysis

#### Training vs. Testing: The Overfitting Signal
*   **Training Set (100% Accuracy):** The model achieved perfect scores across all metrics (Accuracy, F1, ROC-AUC are all 1.0). In Decision Trees, this usually indicates the tree was allowed to grow without limits, creating a complex web of rules that "memorized" every single client in the training database.
*   **Testing Set (95.7% Accuracy):** When exposed to new clients it had never seen, the accuracy dropped slightly to 95.7%.
*   **Insight:** While a drop from 100% to 95.7% indicates overfitting, the test score is still high enough that the model is considered **highly effective**. It has successfully generalized the rules of lending despite the complexity.

#### Reliability (Confusion Matrix Breakdown)
In the testing set of 256 loan applications, the model made only **11 errors**:
*   **False Negatives (5 cases):** The model rejected 5 applicants who were actually eligible.
    *   *Business Impact:* Opportunity cost (lost interest revenue) and potential damage to customer relationships.
*   **False Positives (6 cases):** The model approved 6 applicants who should have been rejected.
    *   *Business Impact:* This is the critical risk category (Potential Default). However, having only 6 bad loans out of 170 approvals is a very healthy risk ratio.

---

### 3. Deep Dive: Feature Importance & Interpretation

The model’s decision-making process is heavily skewed. Below is an interpretation of the drivers behind the approvals:

#### A. The Dominant Factor: Credit Score (77.6%)
*   **Interpretation:** Nearly 80% of the decision weight comes purely from the applicant's credit score.
*   **Insight:** This suggests that the Decision Tree is functioning primarily as a "thresholding" mechanism. If a credit score is above $X$, the loan is likely approved; if below, rejected.
*   **Implication:** While accurate, this raises the question of whether we need a complex Machine Learning model if a simple rules-based engine (e.g., "If Score > 700, Approve") would yield similar results with less computational cost.

#### B. The Secondary Driver: Sentiment (8.3%)
*   **Interpretation:** "Sentiment" likely refers to the Natural Language Processing (NLP) score derived from the applicant's loan statement or interaction logs.
*   **Insight:** It is fascinating that **how** a customer communicates (confidence, positivity, clarity) is the second most important predictor of repayment, ranking higher than income or debt levels. This validates the investment in collecting unstructured text data.

#### C. The "Hidden" Financials (Income, Loan Amount, Job Title)
*   **Observation:** Annual Income (1.0%) and Loan Amount (1.8%) have surprisingly low importance.
*   **Analyst Insight:** Do not mistake low importance for irrelevance. In finance, Credit Score is often highly correlated with Income. Because the Decision Tree splits on Credit Score *first* (capturing the most variance), there is very little "information" left for Income to explain. The Credit Score is effectively acting as a proxy for these financial health markers.

#### D. Compliance Warning: Gender (3.2%) & Age (3.0%)
*   **Risk Flag:** The model is using Gender and Age as top-5 determinants for loan approval.
*   **Compliance Action Required:** In many jurisdictions (e.g., US ECOA standards), using gender as a decision factor for credit is **illegal**. Even though the importance is low (3.2%), its presence means the model is treating men and women differently. **This feature must likely be removed to ensure fair lending compliance.**

---

### 4. Strategic Recommendations

1.  **Compliance Audit (Urgent):** Immediate review of the `Gender` feature. We recommend retraining the model without this variable to ensure we are not violating fair lending laws. Given its low importance, removing it should not significantly impact accuracy.
2.  **Model Pruning:** The "perfect" training score suggests the tree is too deep. We should apply "pruning" (limiting the depth of the tree). This will simplify the rules, make the model more interpretable, and prevent the model from memorizing noise, likely closing the gap between Training and Testing performance.
3.  **Investigate "Sentiment":** Since Sentiment is the #2 predictor, we should analyze *what* specific words or phrases correlate with approval. This could be used to train loan officers on what to look for during interviews.
4.  **Risk Calibration:** The model currently produces slightly more False Positives (Bad Loans approved) than False Negatives. If the bank wants to be more conservative during an economic downturn, we can adjust the probability threshold to minimize those 6 False Positives, even if it means rejecting a few more good customers.

**Conclusion:**
This is a high-performing baseline model. It essentially automates the logic: *"High credit score + Positive interaction = Approval."* With minor adjustments for compliance (removing gender) and optimization (pruning), it is ready for a pilot deployment.

## **Bagging Model (DT with Bagging)**

## **Logistic Regression (LR)**

## **Random Forest (RF)**

## **Gradient Boosting (GBoost)**

## **Extreme Gradient Boosting (XGBoost)**

## **Light Gradient Boosting Machine (LightGBM)**

## **Neural Network (NN)**